### PR TITLE
Add DecIpv4Hop and DecIpv6Hop Elements

### PR DESCRIPTION
**Pull request by eeverett6**
Adds an element to decrement the TTL of an IPv4 packet. Adds a
dependency on smoltcp for packet processing.

This is one of the basic transformation elements from the Click paper,
see #14

### CODE REVIEW

**Comment by eeverett6** [comment]
Wanted to get a trivial networking element in to make sure we're all on the same page. I'm not sure this belongs in the runtime, but I know it's one of the basic elements we'll want to provide.

The warning generated in smoltcp seems to be referenced here, but nowhere else I can find: https://github.com/m-labs/smoltcp/issues/242

**Comment by CollinValley** [comment]
This is basically exactly what I was thinking.  I think we are on the same page here. This could become an element on it's own if you want to resubmit the PR,.  I'm not enamored with the name DecIPTTLElement,  but renaming is easy to do later too.

**Comment by eeverett6** [comment]
@scgruber just want to check with you as well before adding 0BSD as a compatible license for our dependencies.

File: route-rs-runtime/src/element/dec_ip_ttl.rs [hunk 1]
```diff
@@ -0,0 +1,52 @@
+use crate::element::Element;
+use smoltcp::wire::*;
+
+/*
+  DecIPTTL
+  Decrements the TTL of an IPv4 packet
+*/
+#[derive(Default)]
+pub struct DecIPTTLElement {}
```
Comment (moosingin3space) [inline route-rs-runtime/src/element/dec_ip_ttl.rs]: Rust conventions are to avoid having a lot of capital letters next to each other in an identifier. For instance, `DecIpv4HopLimit` is probably a better name. (This is to avoid inconsistency like Java's `URLConnection` and `HttpURLConnection`.)

File: route-rs-runtime/Cargo.toml [hunk 2]
```diff
@@ -8,4 +8,5 @@ license = "MIT"
 [dependencies]
 tokio = "*"
 futures = "*"
-crossbeam = "*"
\ No newline at end of file
+crossbeam = "*"
+smoltcp = "*"
```
Comment (moosingin3space) [inline route-rs-runtime/Cargo.toml]: We should probably pin the version of `smoltcp`.

File: route-rs-runtime/src/element/dec_ip_ttl.rs (hunk 1 above)
Comment (eeverett6) [reply route-rs-runtime/src/element/dec_ip_ttl.rs]: Is DecIpv4Hop sufficient? DecIpv4HopLimitElement just feels a bit wordy

Comment (moosingin3space) [reply route-rs-runtime/src/element/dec_ip_ttl.rs]: How about no `Element` suffix? It's in the `element` module, we don't need to redundantly call it an `Element`.

Comment (eeverett6) [reply route-rs-runtime/src/element/dec_ip_ttl.rs]: Makes sense, updated

File: route-rs-runtime/Cargo.toml (hunk 2 above)
Comment (CollinValley) [reply route-rs-runtime/Cargo.toml]: Good idea, especially since we are only using this for its internal structures.  Once async/await lands and we port tokio I plan on pinning all our dependencies, and maybe make a monthly issue to try and bump the versions when we can.

**Review by CollinValley — DISMISSED** [review]
I approve these changes, but I would like to get a :+1:  from @scgruber about adding the new acceptable license. This one seems compatible with our license schema from my reading but I'd like another pass from someone more versed.

File: route-rs-runtime/src/element/dec_ip_hop.rs [hunk 3]
```diff
@@ -0,0 +1,96 @@
+use crate::element::Element;
+use smoltcp::wire::*;
+
+/*
+  DecIpv4Hop
+  Decrements the TTL of an IPv4 packet
```
Comment (scgruber) [inline route-rs-runtime/src/element/dec_ip_hop.rs]: Conventionally, we should use `///` comments here so they show in documentation. We also don't need to repeat the name in the header since it is shown immediately below.

File: deny.toml [hunk 4]
```diff
@@ -4,6 +4,7 @@ unlicensed = "deny"
 # This lists all of the licenses we will allow in our dependencies. Since route-rs is distributed
 # under the MIT license, we want to make sure all of our dependencies are similarly unrestrictive.
 allow = [
+    "0BSD",
```
Comment (scgruber) [inline deny.toml:7]: LGTM

File: route-rs-runtime/Cargo.toml [hunk 5]
```diff
@@ -8,4 +8,5 @@ license = "MIT"
 [dependencies]
 tokio = "*"
 futures = "*"
-crossbeam = "*"
\ No newline at end of file
+crossbeam = "*"
+smoltcp = "0.5.0"
```
Comment (scgruber) [inline route-rs-runtime/Cargo.toml:12]: For now, let's let our dependencies float unless we have a good reason to lock them.

File: route-rs-runtime/src/element/dec_ip_hop.rs [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+        DecIpv4HopLimit {}
+    }
+}
+
+impl Element for DecIpv4HopLimit {
+    type Input = Ipv4Packet<Vec<u8>>;
+    type Output = Ipv4Packet<Vec<u8>>;
+
+    fn process(&mut self, mut packet: Self::Input) -> Self::Output {
+        packet.set_hop_limit(packet.hop_limit() - 1);
```
Comment (scgruber) [inline route-rs-runtime/src/element/dec_ip_hop.rs]: What happens if the hop limit is currently the minimum allowed? Can we generate an invalid packet by doing this?

File: route-rs-runtime/src/element/dec_ip_hop.rs [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+        packet
+    }
+}
+
+/*
+  DecIpv6Hop
+  Decrements the TTL of an IPv6 packet
```
Comment (scgruber) [inline route-rs-runtime/src/element/dec_ip_hop.rs]: Same as above about comment formatting

File: route-rs-runtime/src/element/dec_ip_hop.rs [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+        DecIpv6HopLimit {}
+    }
+}
+
+impl Element for DecIpv6HopLimit {
+    type Input = Ipv6Packet<Vec<u8>>;
+    type Output = Ipv6Packet<Vec<u8>>;
+
+    fn process(&mut self, mut packet: Self::Input) -> Self::Output {
+        packet.set_hop_limit(packet.hop_limit() - 1);
```
Comment (scgruber) [inline route-rs-runtime/src/element/dec_ip_hop.rs]: Same as above about minimum values

File: route-rs-runtime/src/element/dec_ip_hop.rs [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+        packet
+    }
+}
+
+#[cfg(test)]
+mod tests {
```
Comment (scgruber) [inline route-rs-runtime/src/element/dec_ip_hop.rs:55]: You should add tests for what happens at hop_limit = 0

File: route-rs-runtime/src/element/mod.rs [hunk 10]
```diff
@@ -6,3 +6,6 @@ pub use self::identity_elements::*;
 
 mod transform_elements;
 pub use self::transform_elements::*;
+
+mod dec_ip_hop;
```
Comment (scgruber) [inline route-rs-runtime/src/element/mod.rs:10]: I do like this general structure of putting the IPv4 and IPv6 elements in the same file. It would be even nicer if we could somehow generate both from the same specification, but that's definitely future enhancements, not POC.

**Review by scgruber — CHANGES_REQUESTED** [review]

File: route-rs-runtime/src/element/dec_ip_hop.rs (hunk 3 above)
Comment (CollinValley) [reply route-rs-runtime/src/element/dec_ip_hop.rs]: Seconded, we should begin migrating to docgen comments when it makes sense

File: route-rs-runtime/src/element/dec_ip_hop.rs (hunk 6 above)
Comment (CollinValley) [reply route-rs-runtime/src/element/dec_ip_hop.rs]: One quick fix is to just never set this value below zero, and then leave it up to a classify element later to check ttl is greater than zero? The click codebase will be a decent guide

File: route-rs-runtime/src/element/dec_ip_hop.rs (hunk 3 above)
Comment (eeverett6) [reply route-rs-runtime/src/element/dec_ip_hop.rs]: Done

File: route-rs-runtime/src/element/dec_ip_hop.rs (hunk 6 above)
Comment (eeverett6) [reply route-rs-runtime/src/element/dec_ip_hop.rs]: I'm limiting this to 0 and passing the packet along unchanged. When we have more support for ICMP errors in the future we can generate an ICMP error instead.

**Review by scgruber — APPROVED** [review]
